src/Core/Stock/StockManager.php

```diff
@@ -174,9 +174,99 @@ public function updateQuantity($product, $id_product_attribute, $delta_quantity,
             )
         );
 
+        if ($this->checkIfMustSendLowStockAlert($product, $id_product_attribute, $stockAvailable->quantity)) {
+            $this->sendLowStockAlert($product, $id_product_attribute, $stockAvailable->quantity);
+        }
+
         $cacheManager->clean('StockAvailable::getQuantityAvailableByProduct_'.(int)$product->id.'*');
     }
 
+    protected function checkIfMustSendLowStockAlert($product, $id_product_attribute, $newQuantity)
+    {
+        if (!\Configuration::get('PS_STOCK_MANAGEMENT')) {
+            return false;
+        }
+
+        // Do not send mail if multiples product are created / imported.
+        if (defined('PS_MASS_PRODUCT_CREATION')) {
+            return false;
+        }
+
+        $product_has_attributes = $product->hasAttributes();
+        if ($product_has_attributes && $id_product_attribute) {
+            $combination = new \Combination($id_product_attribute);
+            if ($combination->low_stock_alert
+                && (int) $combination->low_stock_threshold > 0
+                && $newQuantity <= (int) $combination->low_stock_threshold
+            ) {
+                return true;
+            }
+        } elseif ($product_has_attributes && !$id_product_attribute) {
+            if ($product->low_stock_alert
+                && (int) $product->low_stock_threshold > 0
+                && $newQuantity <= (int) $product->low_stock_threshold
+            ) {
+                return true;
+            }
+        }
```

Comment:
I bet Codacy is grimacing about this if/else block. I don't like it either.
I'd go with : 
```php
if ($product->hasAttributes()) {
    $stuffToTest = (bool)$id_product_attribute 
        ? new \Combination($id_product_attribute) 
        : $product;
    if ($stuffToTest->low_stock_alert
        && (int) $stuffToTest->low_stock_threshold > 0
        && $newQuantity <= (int) $stuffToTest->low_stock_threshold
    ) {
        return true;
    }
}
```
And to go further, I would even create a method testing if "stuff" should trigger an alert (the method would test `low_stock_alert`, `low_stock_threshold` and quantity on "stuff")

Reply:
I would not factorize here since the objects $product and $combination do not implement same interface : it would inject code refacto hardness
eg: rename Product::$low_stock_alert without renaming Combination::$low_stock_alert (allowed since not implementing same interface)

I understand the code is hard to read, I split it into 2 distinct methods ;)